### PR TITLE
[ML] add zero_shot_classification task for BERT nlp models

### PR DESCRIPTION
**Pull request by benwtrent**
Zero-Shot classification allows for text classification tasks without a pre-trained collection of target labels.

This is achieved through models trained on the Multi-Genre Natural Language Inference (MNLI) dataset. This dataset pairs  text sequences with "entailment" clauses. An example could be:

"Throughout all of history, man kind has shown itself resourceful, yet astoundingly short-sighted" could have been paired with the entailment clauses: ["This example is history", "This example is sociology"...]. 

This training set combined with the attention and semantic knowledge in modern day NLP models (BERT, BART, etc.) affords a powerful tool for ad-hoc text classification.

See https://arxiv.org/abs/1909.00161 for a deeper explanation of the MNLI training and how zero-shot works. 

The zeroshot classification task is configured as follows:
```js
{
   // <snip> model configuration </snip>
  "inference_config" : {
    "zero_shot_classification": {
      "classification_labels": ["entailment", "neutral", "contradiction"], // <1>
      "labels": ["sad", "glad", "mad", "rad"], // <2>
      "multi_label": false, // <3>
      "hypothesis_template": "This example is {}.", // <4>
      "tokenization": { /*<snip> tokenization configuration </snip>*/}
    }
  }
}
```
* <1> For all zero_shot models, there returns 3 particular labels when classification the target sequence. "entailment" is the positive case, "neutral" the case where the sequence isn't positive or negative, and "contradiction" is the negative case
* <2> This is an optional parameter for the default zero_shot labels to attempt to classify
* <3> When returning the probabilities, should the results assume there is only one true label or multiple true labels
* <4> The hypothesis template when tokenizing the labels. When combining with `sad` the sequence looks like `This example is sad.`

For inference in a pipeline one may provide label updates:
```js
{
  //<snip> pipeline definition </snip>
  "processors": [
    //<snip> other processors </snip>
    {
      "inference": {
        // <snip> general configuration </snip>
        "inference_config": {
          "zero_shot_classification": {
             "labels": ["humanities", "science", "mathematics", "technology"], // <1>
             "multi_label": true // <2>
          }
        }
      }
    }
    //<snip> other processors </snip>
  ]
}
```
* <1> The `labels` we care about, these replace the default ones if they exist. 
* <2> Should the results allow multiple true labels

Similarly one may provide label changes against the `_infer` endpoint
```js
{
   "docs":[{ "text_field": "This is a very happy person"}],
   "inference_config":{"zero_shot_classification":{"labels": ["glad", "sad", "bad", "rad"], "multi_label": false}}
}
```

### CODE REVIEW

**Comment by benwtrent** [comment]
run elasticsearch-ci/part-2

**Review by davidkyle — COMMENTED** [review]
Looks good, only some minor comments

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NlpConfigUpdate.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+import org.elasticsearch.common.xcontent.ParseField;
+
+public abstract class NlpConfigUpdate implements InferenceConfigUpdate {
+
+    static ParseField CLASSIFICATION_LABELS = new ParseField("classification_labels");
+
+    @Override
+    public InferenceConfig toConfig() {
+        throw new UnsupportedOperationException("cannot serialize to nodes before 7.8");
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NlpConfigUpdate.java:18]: @davidkyle we cannot create an NLP inference config from an update. In a separate PR I am gonna remove this check and this method as it is not used in master (so, this is an intermediate change really).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ZeroShotClassificationConfig.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                REQUIRED_CLASSIFICATION_LABELS
+            );
+        }
+        List<String> badLabels = classificationLabels.stream()
+            .map(s -> s.toLowerCase(Locale.ROOT))
+            .filter(c -> REQUIRED_CLASSIFICATION_LABELS.contains(c) == false)
+            .collect(Collectors.toList());
+        if (badLabels.isEmpty() == false) {
+            throw ExceptionsHelper.badRequestException(
+                "[{}] must contain exactly the three values {}",
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ZeroShotClassificationConfig.java:124]: ++

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ZeroShotClassificationConfig.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+                CLASSIFICATION_LABELS.getPreferredName(),
+                REQUIRED_CLASSIFICATION_LABELS
+            );
+        }
+        this.vocabularyConfig = Optional.ofNullable(vocabularyConfig)
+            .orElse(new VocabularyConfig(InferenceIndexConstants.nativeDefinitionStore()));
+        this.tokenization = tokenization == null ? Tokenization.createDefault() : tokenization;
+        this.isMultiLabel = isMultiLabel != null && isMultiLabel;
+        this.hypothesisTemplate = Optional.ofNullable(hypothesisTemplate).orElse(DEFAULT_HYPOTHESIS_TEMPLATE);
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ZeroShotClassificationConfig.java:133]: > rename labels to hypothesisLabels

I don't think anybody calls them that. 

another good option might be `target_labels`?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ZeroShotClassificationConfigUpdate.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    public static class Builder implements InferenceConfigUpdate.Builder<
+        ZeroShotClassificationConfigUpdate.Builder,
+        ZeroShotClassificationConfigUpdate
+        > {
+        private List<String> labels;
+        private Boolean isMultiLabel;
+
+        @Override
+        public ZeroShotClassificationConfigUpdate.Builder setResultsField(String resultsField) {
+            throw new IllegalArgumentException();
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ZeroShotClassificationConfigUpdate.java:184]: Well, we don't even have a `results` field in the original NLP configs. 

I think this is a larger discussion around unifying the NLP Configs with the classification/regression configs.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ZeroShotClassificationConfigTests.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    @Override
+    protected ZeroShotClassificationConfig mutateInstanceForVersion(ZeroShotClassificationConfig instance, Version version) {
+        return instance;
+    }
+
+    public static ZeroShotClassificationConfig createRandom() {
+        return new ZeroShotClassificationConfig(
+            randomFrom(List.of("entailment", "neutral", "contradiction"), List.of("entailment", "neutral", "contradiction")),
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ZeroShotClassificationConfigTests.java:53]: yeah, entailment and contradiction should be flipped, changing.

File: docs/reference/ml/df-analytics/apis/get-trained-models.asciidoc [hunk 6]
```diff
@@ -414,6 +414,68 @@ include::{es-repo-dir}/ml/ml-shared.asciidoc[tag=inference-config-nlp-tokenizati
 (Optional, integer)
 include::{es-repo-dir}/ml/ml-shared.asciidoc[tag=inference-config-nlp-tokenization-bert-max-sequence-length]
 
+`with_special_tokens`::::
```
Comment (benwtrent) [inline docs/reference/ml/df-analytics/apis/get-trained-models.asciidoc:417]: @szabosteve mind taking a look at the doc changes?

File: docs/reference/ml/ml-shared.asciidoc [hunk 7]
```diff
@@ -998,6 +998,46 @@ prediction. Defaults to the `results_field` value of the {dfanalytics-job} that
...
+If consistently classifying the same labels, it may be better to use an optimized
+text classification model.
+end::inference-config-zero-shot-classification[]
+
+tag::inference-config-zero-shot-classification-classification-labels[]
+The classification labels used during the zero-shot classification. Classification
+labels must not be empty or null and only set at model creation. They must be all three
+of ["entailment", "neutral", "contradiction"].
+
+NOTE: this is NOT the same as `labels` which are the values that zero-shot is attempting to
```
Comment (szabosteve) [inline docs/reference/ml/ml-shared.asciidoc:1016]:
```suggestion
NOTE: This is NOT the same as `labels` which are the values that zero-shot is attempting to
```

File: docs/reference/ml/ml-shared.asciidoc [hunk 8]
```diff
@@ -998,6 +998,46 @@ prediction. Defaults to the `results_field` value of the {dfanalytics-job} that
...
+      classify.
+end::inference-config-zero-shot-classification-classification-labels[]
+
+tag::inference-config-zero-shot-classification-hypothsis-template[]
+This is the template used when tokenizing the sequences for classification.
+
+The labels replace the `{}` value in the text. The default value is:
+`This This example is {}.`
```
Comment (szabosteve) [inline docs/reference/ml/ml-shared.asciidoc:1024]:
```suggestion
`This example is {}.`
```

**Review by szabosteve — APPROVED** [review]
Documentation LGTM! Thanks for writing the content! ✍️ I left three suggestions. [This one](https://github.com/elastic/elasticsearch/pull/77799#discussion_r717261395) will make the docs CI pass.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        if (nlpConfig instanceof ZeroShotClassificationConfig) {
+            ZeroShotClassificationConfig zeroShotConfig = (ZeroShotClassificationConfig) nlpConfig;
+            labels = zeroShotConfig.getLabels().toArray(new String[0]);
+        } else {
+            labels = this.labels;
+        }
+        if (this.labels == null || this.labels.length == 0) {
+            throw ExceptionsHelper.badRequestException("zero_shot_classification requires non-empty [labels]");
+        }
+        return new RequestBuilder(tokenizer,  labels, hypothesisTemplate);
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java:75]:
```suggestion
        return new RequestBuilder(tokenizer, labels, hypothesisTemplate);
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                int v = 0;
+                for (double[] vals : pyTorchResult.getInferenceResult()[0]) {
+                    if (vals.length != 3) {
+                        return new WarningInferenceResults(
+                            "Expected exactly [{}] values in inner zero shot classification result; got [{}]",
+                            3,
+                            vals.length
+                        );
+                    }
+                    // assumes entailment is `0`, it COULD be `2`
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java:179]: is this comment out of date? It seems to be referring to a hard-coded vale rather than the use of `entailmentPos`

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+            // checked. Implement for all in batch
+            if (pyTorchResult.getInferenceResult()[0].length != labels.length) {
+                return new WarningInferenceResults(
+                    "Expected exactly [{}] values in zero shot classification result; got [{}]",
+                    labels.length,
+                    pyTorchResult.getInferenceResult().length
+                );
+            }
+            final double[] normalizedScores;
+            if (isMultiLabel) {
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java:152]: `isMultiLabel` is just about how the results are interpreted? When true the probability of entailment as opposed to contradiction for each label is returned. When false it is the probability of each label being entailment

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ZeroShotClassificationConfigUpdate.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ZeroShotClassificationConfigUpdate.java:184]: It should really be part of the inference processor config. Looking at the code this function appears to exist for use by `InferencePipelineAggregationBuilder`. ++ to revisiting this and simplifying the code.

File: docs/reference/ml/ml-shared.asciidoc [hunk 12]
```diff
@@ -998,6 +998,46 @@ prediction. Defaults to the `results_field` value of the {dfanalytics-job} that
 used to train the model, which defaults to `<dependent_variable>_prediction`.
 end::inference-config-results-field-processor[]
 
+tag::inference-config-zero-shot-classification[]
+Configures a zero-shot classification task. Zero-shot classification allows for
+text classification to occur without pre-determined labels. At inference time,
+it is possible to adjust the labels to classify. This makes this type of model
+and task exceptionally flexible.
+
+If consistently classifying the same labels, it may be better to use an optimized
```
[the remark this answers is not in the document]
Comment (szabosteve) [reply docs/reference/ml/ml-shared.asciidoc:1007]: If you accept this suggestion, please also change the indefinite article to `a` from `an`.

Comment (benwtrent) [inline docs/reference/ml/ml-shared.asciidoc:1007]:
```suggestion
If consistently classifying the same labels, it may be better to use a fine turned
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java (hunk 11 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java:152]: `isMultiLabel` is basically:

 - When `true` its `softmax` of individual entailment vs contradiction (probs don't sum to 1.0)
 - When `false` its `softmax` of all entailments (probs sum to 1.0)

The docs already state you use it when you could have more than one true label. Which is exactly what we use it for.

I would rather not talk about softmax, entailment, etc.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java (hunk 10 above)
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java:179]:
```suggestion
```

**Review by davidkyle — APPROVED** [review]
LGTM
